heap/heap.go

```
@@ -14,43 +16,57 @@ import (
 // but no more often than every minTimeDiff seconds
 // any errors will be sent to the errors channel
 type Heap struct {
-	path        string
-	threshold   int
-	minTimeDiff int
-	checkEvery  time.Duration
-	lastUnix    int64
-	Errors      chan error
+	cfg           Config
+	lastTriggered time.Time
+	Errors        chan error
+
+	proc *procfs.Proc
+}
+
+// Config is the config for triggering profile
+type Config struct {
+	Path           string
+	AllocThreshold int
+	RSSThreshold   int
+	MinTimeDiff    time.Duration
+	CheckEvery     time.Duration
 }
 
 // New creates a new Heap trigger. use a nil channel if you don't care about any errors
-func New(path string, threshold, minTimeDiff int, checkEvery time.Duration, errors chan error) (*Heap, error) {
+func New(cfg Config, errors chan error) (*Heap, error) {
 	heap := Heap{
-		path,
-		threshold,
-		minTimeDiff,
-		checkEvery,
-		int64(0),
-		errors,
+		cfg:           cfg,
+		lastTriggered: time.Now().Add(-cfg.MinTimeDiff),
+		Errors:        errors,
+	}
+	proc, err := procfs.Self()
+	if err != nil {
+		heap.logError(err)
+	} else {
+		heap.proc = &proc
 	}
+
 	return &heap, nil
 }
 
 func (heap Heap) logError(err error) {
 	if heap.Errors != nil {
-		heap.Errors <- err
+		select {
+		case heap.Errors <- err:
```

why silently drop errors? it's the caller responsibility to read the errors (if a non-nil chan was provided)